core/domain/question_jobs_one_off.py

```diff
@@ -120,6 +123,62 @@ def reduce(key, values):
         yield (key, values)
 
 
+class FixQuestionImagesStorageOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """This job is used to ensure that all question related images are present
+    in the correct storage path.
+    """
+
+    _IMAGE_COPIED = 'question_image_copied'
+    _DELETED_KEY = 'question_deleted'
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [question_models.QuestionModel]
+
+    @staticmethod
+    def map(item):
+        if item.deleted:
+            yield (FixQuestionImagesStorageOneOffJob._DELETED_KEY, 1)
+            return
+
+        question = question_services.get_question_by_id(item.id, strict=False)
+        html_list = question.question_state_data.get_all_html_content_strings()
+        image_filenames = html_cleaner.get_image_filenames_from_html_strings(
+            html_list)
+        file_system_class = fs_services.get_entity_file_system_class()
+        question_fs = fs_domain.AbstractFileSystem(file_system_class(
+            feconf.ENTITY_TYPE_QUESTION, question.id))
+        success_count = 0
+        # For each image filename, check if it exists in the correct path. If
+        # not, copy the image file the correct path else continue.
+        for image_filename in image_filenames:
+            if not question_fs.isfile('image/%s' % image_filename):
+                for skill_id in question.linked_skill_ids:
+                    skill_fs = fs_domain.AbstractFileSystem(file_system_class(
+                        feconf.ENTITY_TYPE_SKILL, skill_id))
+                    if skill_fs.isfile('image/%s' % image_filename):
+                        fs_services.copy_images(
+                            feconf.ENTITY_TYPE_SKILL, skill_id,
+                            feconf.ENTITY_TYPE_QUESTION, question.id,
+                            [image_filename])
+                        success_count += 1
+                        break
```

Comment:
Why break? A question can be linked to multiple skills.

Reply:
Yup but copying one file corresponding to the filename is sufficient, this is why I'm breaking out of the loop if we find the filename in any of the skills the question is linked to.

Reply:
Oh, missed that it was iterating on image filenames, sg!